### PR TITLE
Support Promise execution in the sandbox.

### PR DESCRIPTION
**Pull request by kevinswiber**
Promise.prototype.then arguments are executed in a microtask queue,
pushed to the end of the current event loop tick. The async
lifetime tracking in the sandbox did not factor this into
consideration. Certain completion functions needed to be deferred
with setImmediate(), which will add to the task queue and
execute after the current microtask queue drains.

Calls to global timers also needed to be wrapped in a scoped context.
Due to the setImmediate() delay, the bridge may disconnect when the
context is disposed before the completion timer fires. Disconnecting
the bridge removes all globals. This prevents reference errors from
happening on the final tick.

Original behavior of execution.return.async remains.  It will only
be set to `true` if there's a pending timer that hasn't been
previously cleared.

A code example that will show the current issue of early exit:

```js
const foo = () => new Promise((resolve) => {
  console.log('foo');
  setTimeout(resolve, 200);
});

const bar = () => new Promise((resolve) => {
  console.log('bar');
  resolve();
});

const baz = () => new Promise((resolve) => {
  console.log('baz');
  resolve();
});

Promise.resolve()
  .then(() => foo())
  .then(() => bar())
  .then(() => baz());
```

The issue occurs when async tracking happens within the microtask queue.  Without these changes, the sandbox believes all async functions are done and exits.  Using `setImmediate` in a couple of places defers that check to the beginning of the next tick, after the current tick's microtask queue is drained.

### CODE REVIEW

**Comment by kevinswiber** [comment]
@codenirvana I think all the sandbox version checking might be a red herring here.  We're hitting an issue on the `context.dispose()` tests.  These changes allow the current tick to finish.  The current event loop tick has to finish in order to let the microtask queue used by Promises execute and drain.  However, when a `PostmanSandbox.prototype.disconnect` happens, the UVM bridge is removing all the globals from the context.  Therefore, given this code:

```js
console.log('started');

setTimeout(function () {
    console.log('timeout');
}, 50);
```
And a `setTimeout(context.dispose.bind(context), 1)`...

The behavior prior to this PR might halt execution of the script after the `console.log` statement.  With these changes, the current loop completes, even with the `dispose` call happening after 1ms.  Since `dispose` removes all globals, we get an error of `Uncaught ReferenceError: setTimeout is not defined`.

We could:
* Try to defer a `dispose` call until the end of the current loop, or...
* Swallow script errors that occur after a `dispose` completes.

Thoughts?

**Comment by kevinswiber** [comment]
@codenirvana I think I found a way around it: using wrapped timer functions that aren't in globals.  There's now support for legacy sandboxes.

**Comment by kevinswiber** [comment]
Note that this PR does not address unhandled Promise rejections.  That will need to be a separate PR to the https://github.com/postmanlabs/uvm repo.

File: lib/sandbox/timers.js [hunk 1]
```diff
@@ -158,11 +157,15 @@ function Timerz (delegations, onError, onAnyTimerStart, onAllTimerEnd) {
     computeTimerEvents = function (increment, clearing) {
         increment && (pending += increment);
 
-        if (pending === 0 && computeTimerEvents.started) {
-            !clearing && (typeof onAllTimerEnd === FUNCTION) && onAllTimerEnd();
-            computeTimerEvents.started = false;
+        function maybeEndAllTimers () {
+            if (pending === 0 && computeTimerEvents.started) {
```
Comment (codenirvana) [inline lib/sandbox/timers.js:161]: This looks like a duplicate check from line 167.

Comment (kevinswiber) [reply lib/sandbox/timers.js:161]: @codenirvana The first check is "only queue this if `pending` is currently at 0."  The second check is "only actually run this if `pending` is still at 0."  This is the one we really need.  The first check just prevents needless queueing.

**Review by codenirvana — APPROVED** [review]

**Pull request merged**